loopd/daemon.go

```diff
@@ -159,100 +181,258 @@ func (d *Daemon) Run() error {
+	// and testnet.
+	if d.cfg.SwapServer == "" {
+		// TODO(wilmer): Use onion service addresses when proxy is
+		// active.
+		switch d.cfg.Network {
+		case "mainnet":
+			d.cfg.SwapServer = mainnetServer
+		case "testnet":
+			d.cfg.SwapServer = testnetServer
+		default:
+			return errors.New("no swap server address specified")
+		}
+	}
+
+	log.Infof("Swap server address: %v", d.cfg.SwapServer)
+
+	// Create an instance of the loop client library.
+	swapclient, clientCleanup, err := getClient(d.cfg, &d.lnd.LndServices)
+	if err != nil {
+		return err
+	}
+	d.clientCleanup = clientCleanup
+
+	// Both the client RPC server and and the swap server client should
+	// stop on main context cancel. So we create it early and pass it down.
+	d.mainCtx, d.mainCtxCancel = context.WithCancel(context.Background())
+
+	// Now finally fully initialize the swap client RPC server instance.
+	d.swapClientServer = swapClientServer{
+		impl:        swapclient,
+		lnd:         &d.lnd.LndServices,
+		swaps:       make(map[lntypes.Hash]loop.SwapInfo),
+		subscribers: make(map[int]chan<- interface{}),
+		statusChan:  make(chan loop.SwapInfo),
+		mainCtx:     d.mainCtx,
+	}
+
+	// Retrieve all currently existing swaps from the database.
+	swapsList, err := d.impl.FetchSwaps()
+	if err != nil {
+		return err
+	}
+
+	for _, s := range swapsList {
+		d.swaps[s.SwapHash] = *s
+	}
+
+	// Start the main event and error handlers, now that we have prepared
+	// everything they need to work.
+	d.startHandlers()
+
+	return nil
+}
 
+// startHandlers starts all goroutines that are tasked with handling swap and
+// error events.
+func (d *Daemon) startHandlers() {
 	// Start the swap client itself.
-	wg.Add(1)
+	d.wg.Add(1)
 	go func() {
-		defer wg.Done()
+		defer d.wg.Done()
 
 		log.Infof("Starting swap client")
-		err := swapClient.Run(mainCtx, d.statusChan)
+		err := d.impl.Run(d.mainCtx, d.statusChan)
 		if err != nil {
-			log.Error(err)
+			// Notify the main error handler goroutine that
+			// we exited unexpectedly here. We don't have to
+			// worry about blocking as the internal error
+			// channel is sufficiently buffered.
+			d.internalErrChan <- err
 		}
 		log.Infof("Swap client stopped")
-
-		log.Infof("Stopping gRPC server")
-		grpcServer.Stop()
-
-		cancel()
 	}()
 
 	// Start a goroutine that broadcasts swap updates to clients.
-	wg.Add(1)
+	d.wg.Add(1)
 	go func() {
-		defer wg.Done()
+		defer d.wg.Done()
 
 		log.Infof("Waiting for updates")
-		d.processStatusUpdates(mainCtx)
+		d.processStatusUpdates(d.mainCtx)
 	}()
 
-	// Start the grpc server.
-	wg.Add(1)
+	// Last, start our internal error handler. This will return at most one
+	// error on the main error channel to inform the caller that something
+	// went wrong.
+	d.wg.Add(1)
 	go func() {
-		defer wg.Done()
+		defer d.wg.Done()
 
-		log.Infof("RPC server listening on %s", grpcListener.Addr())
+		defer func() {
+			err := d.Stop()
```

Comment:
Does this work? Waiting (in Stop) on the wait group inside a goroutine that is added to the wait group?

Reply:
Yes, because of the `defer` above. The order of the defers is kept. I added a comment to clarify.

Reply:
Hmm... I thought it would because of the order of the `defer`s. But you are right, it didn't work.
Solved it with a helper function `ensureShutdown`. All my manual test cases now work correctly.